From 2ade027daddda6e74f61e1fd87b8192dae04e3ae Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 11 May 2023 19:42:04 +0200
Subject: [PATCH 0751/1695] clk: stm32mp1: add dsb in clock driver

Add memory barriers in RCC clock driver to ensure the system is in the
expected state when requests are proceeded by RCC. No pending register
operation before disabling the clocks and return to caller only when
clock is enabled, so before any accesses to the clocked devices.

As the registers are mapped as device memory (shareable, bufferable),
the order of operation is guarantee only at outer shareable limit
and not on each device, for example when they are not on the same bus.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 203c13d0a93dbeeaba8d27e14f97925ebbe91bce clk: stm32mp1: add dsb in clock driver]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I4ec28cc6e190af6fc7b2c6bc92d7b1557dd51f48
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/305819
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/373109
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
Reviewed-by: Gabriel FERNANDEZ <gabriel.fernandez@foss.st.com>
---
 core/drivers/clk/clk-stm32-core.c |  5 +++++
 core/drivers/clk/clk-stm32mp13.c  |  3 +++
 core/drivers/clk/clk-stm32mp15.c  | 19 +++++++++++++++++--
 3 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/core/drivers/clk/clk-stm32-core.c b/core/drivers/clk/clk-stm32-core.c
index 98922aea9..d2a716d83 100644
--- a/core/drivers/clk/clk-stm32-core.c
+++ b/core/drivers/clk/clk-stm32-core.c
@@ -73,7 +73,12 @@ void stm32_gate_endisable(uint16_t gate_id, bool enable)
 			io_write32(addr, BIT(gate->bit_idx));
 		else
 			io_setbits32_stm32shregs(addr, BIT(gate->bit_idx));
+		/* Make sure the clock is enabled before returning to caller */
+		dsb();
 	} else {
+		/* Waiting pending operation before disabling clock */
+		dsb();
+
 		if (gate->set_clr)
 			io_write32(addr + RCC_MP_ENCLRR_OFFSET,
 				   BIT(gate->bit_idx));
diff --git a/core/drivers/clk/clk-stm32mp13.c b/core/drivers/clk/clk-stm32mp13.c
index 4b00c9497..635f54d45 100644
--- a/core/drivers/clk/clk-stm32mp13.c
+++ b/core/drivers/clk/clk-stm32mp13.c
@@ -3512,5 +3512,8 @@ void stm32_reset_system(void)
 {
 	vaddr_t rcc = stm32_rcc_base();
 
+	/* Make sure the pending operation are visible */
+	dsb();
+
 	io_write32(rcc + RCC_MP_GRSTCSETR, RCC_MP_GRSTCSETR_MPSYSRST);
 }
diff --git a/core/drivers/clk/clk-stm32mp15.c b/core/drivers/clk/clk-stm32mp15.c
index 109076463..29ad2320d 100644
--- a/core/drivers/clk/clk-stm32mp15.c
+++ b/core/drivers/clk/clk-stm32mp15.c
@@ -1954,17 +1954,25 @@ DECLARE_KEEP_PAGER(clk_op_compute_rate);
 
 static TEE_Result clk_op_enable(struct clk *clk)
 {
-	if (clk_is_gate(clk))
+	if (clk_is_gate(clk)) {
 		__clk_enable(clk_to_gate_ref(clk));
 
+		/* Make sure the clock is enabled before returning to caller */
+		dsb();
+	}
+
 	return TEE_SUCCESS;
 }
 DECLARE_KEEP_PAGER(clk_op_enable);
 
 static void clk_op_disable(struct clk *clk)
 {
-	if (clk_is_gate(clk))
+	if (clk_is_gate(clk)) {
+		/* Make sure the previous operations are visible */
+		dsb();
+
 		__clk_disable(clk_to_gate_ref(clk));
+	}
 }
 DECLARE_KEEP_PAGER(clk_op_disable);
 
@@ -4224,12 +4232,16 @@ static TEE_Result stm32_clock_pm(enum pm_op op, unsigned int pm_hint __unused,
 				 const struct pm_callback_handle *hdl __unused)
 {
 	if (op == PM_OP_SUSPEND) {
+		/* Make sure the pending operations are visible */
+		dsb();
 		stm32_clock_suspend();
 	} else {
 #ifdef CFG_STM32_CPU_OPP
 		pll1_config_from_opp_khz(current_opp_khz);
 #endif
 		stm32_clock_resume();
+		/* Make sure the clock are restored */
+		dsb();
 	}
 
 	return TEE_SUCCESS;
@@ -4240,5 +4252,8 @@ void stm32_reset_system(void)
 {
 	vaddr_t rcc = stm32_rcc_base();
 
+	/* Make sure the pending operations are visible */
+	dsb();
+
 	io_write32(rcc + RCC_MP_GRSTCSETR, RCC_MP_GRSTCSETR_MPSYSRST);
 }
-- 
2.47.2

